File: Sources/TranscriptionKit/DeepgramBatchProvider.swift
import Foundation

// MARK: - DeepgramBatchProvider

/// `BatchTranscriptionProvider` adapter over Deepgram's REST endpoint
/// `https://api.deepgram.com/v1/listen`. Used by RecorderState when the user
/// picks Deepgram in Settings — uploads the whole `audio.m4a` after Stop and
/// gets back a single transcript with word-level timestamps.
///
/// This is intentionally separate from the streaming `DeepgramProvider`: same
/// vendor, same auth, completely different I/O shape. The streaming provider
/// stays for v1.1 when capture grows a PCM tee.
public final class DeepgramBatchProvider: BatchTranscriptionProvider, Sendable {

    public typealias HTTPClient = @Sendable (URLRequest) async throws -> (Data, URLResponse)

    private let apiKey: String
    private let endpoint: URL
    private let model: String
    private let httpClient: HTTPClient

    public init(
        apiKey: String,
        endpoint: URL = URL(string: "https://api.deepgram.com/v1/listen")!,
        model: String = "nova-2",
        httpClient: @escaping HTTPClient = DeepgramBatchProvider.defaultHTTPClient
    ) {
        self.apiKey = apiKey
        self.endpoint = endpoint
        self.model = model
        self.httpClient = httpClient
    }

    public func transcribe(audioFile: URL, config: TranscriptionConfig) async throws -> BatchTranscriptResult {
        // Build URL: /v1/listen?model=…&punctuate=true[&language=…]
        var components = URLComponents(url: endpoint, resolvingAgainstBaseURL: false)!
        var items = [
            URLQueryItem(name: "model", value: model),
            URLQueryItem(name: "punctuate", value: "true"),
            URLQueryItem(name: "smart_format", value: "true"),
        ]
        if let language = config.language, !language.isEmpty, language.lowercased() != "auto" {
            items.append(URLQueryItem(name: "language", value: language))
        }
        components.queryItems = items
        guard let url = components.url else {
            throw DeepgramBatchError.invalidEndpoint
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Token \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("audio/m4a", forHTTPHeaderField: "Content-Type")

        let audioData: Data
        do {
            audioData = try Data(contentsOf: audioFile)
        } catch {
            throw DeepgramBatchError.audioReadFailed(underlying: error)
        }
        request.httpBody = audioData

        let (data, response) = try await httpClient(request)
        guard let http = response as? HTTPURLResponse else {
            throw DeepgramBatchError.unexpectedResponse(status: -1, body: nil)
        }
        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8)
            throw DeepgramBatchError.unexpectedResponse(status: http.statusCode, body: body)
        }

        return try Self.parse(data: data)
    }

    // MARK: - Defaults

    public static let defaultHTTPClient: HTTPClient = { request in
        try await URLSession.shared.data(for: request)
    }

    // MARK: - Parsing

    /// Parse Deepgram batch JSON into a `BatchTranscriptResult`.
    /// Schema: `results.channels[0].alternatives[0]` carries the transcript
    /// + word-level timing; `metadata.duration` carries the audio length.
    static func parse(data: Data) throws -> BatchTranscriptResult {
        struct Envelope: Decodable {
            let results: ResultsBlock
            let metadata: MetadataBlock?
        }
        struct ResultsBlock: Decodable {
            let channels: [Channel]
        }
        struct Channel: Decodable {
            let alternatives: [Alternative]
            let detected_language: String?
        }
        struct Alternative: Decodable {
            let transcript: String
            let words: [Word]?
        }
        struct Word: Decodable {
            let start: Double
            let end: Double
            let punctuated_word: String?
            let word: String
        }
        struct MetadataBlock: Decodable {
            let duration: Double?
        }

        let env: Envelope
        do {
            env = try JSONDecoder().decode(Envelope.self, from: data)
        } catch {
            throw DeepgramBatchError.parseFailed(underlying: error)
        }

        guard let channel = env.results.channels.first,
              let alt = channel.alternatives.first else {
            throw DeepgramBatchError.noTranscript
        }

        let duration = env.metadata?.duration ?? 0
        let language = channel.detected_language

        // Group words into ~5 s "segments" so the Library transcript view stays
        // readable — Deepgram returns one word per timestamp, which would make
        // 1000-row tables for a 20-min meeting.
        let words = alt.words ?? []
        var segments: [TranscriptSegment] = []
        if words.isEmpty {
            // No word-level timing: synthesize a single segment.
            segments.append(TranscriptSegment(
                start: 0,
                end: duration,
                text: alt.transcript,
                confidence: 1.0,
                isFinal: true,
                speaker: nil
            ))
        } else {
            var cursor = 0
            while cursor < words.count {
                let segmentStart = words[cursor].start
                let cutoff = segmentStart + 5.0
                var endIndex = cursor
                while endIndex < words.count && words[endIndex].end <= cutoff {
                    endIndex += 1
                }
                if endIndex == cursor { endIndex = cursor + 1 }
                let chunk = words[cursor..<endIndex]
                let text = chunk
                    .map { $0.punctuated_word ?? $0.word }
                    .joined(separator: " ")
                let segmentEnd = chunk.last?.end ?? segmentStart
                segments.append(TranscriptSegment(
                    start: segmentStart,
                    end: segmentEnd,
                    text: text,
                    confidence: 1.0,
                    isFinal: true,
                    speaker: nil
                ))
                cursor = endIndex
            }
        }

        return BatchTranscriptResult(
            language: language,
            duration: duration,
            segments: segments,
            text: alt.transcript
        )
    }
}

// MARK: - Errors

public enum DeepgramBatchError: Error, Sendable {
    case invalidEndpoint
    case audioReadFailed(underlying: Error)
    case unexpectedResponse(status: Int, body: String?)
    case parseFailed(underlying: Error)
    case noTranscript
}

File: windows/src/KosmoNotes.Providers/DeepgramBatchProvider.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using KosmoNotes.Core.Models;
using KosmoNotes.Core.Transcription;
using KosmoNotes.Providers.Internal;

namespace KosmoNotes.Providers;

/// <summary>
/// <see cref="IBatchTranscriptionProvider"/> over Deepgram's REST endpoint
/// <c>https://api.deepgram.com/v1/listen</c>. Mirrors the Swift
/// <c>DeepgramBatchProvider</c>.
/// </summary>
/// <remarks>
/// Uploads the whole audio file after stop and gets back a single transcript
/// with word-level timestamps. Words are grouped into ~5 s segments so the
/// Library transcript view stays readable.
/// </remarks>
public sealed class DeepgramBatchProvider : IBatchTranscriptionProvider
{
    /// <summary>The default batch listen endpoint.</summary>
    public static readonly Uri DefaultEndpoint =
        new("https://api.deepgram.com/v1/listen");

    /// <summary>The model used by default.</summary>
    public const string DefaultModel = "nova-2";

    private readonly string _apiKey;
    private readonly Uri _endpoint;
    private readonly string _model;
    private readonly HttpClient _httpClient;

    /// <summary>Create a provider using the given API key.</summary>
    /// <param name="apiKey">Deepgram API key (sent as <c>Authorization: Token …</c>).</param>
    /// <param name="endpoint">Optional endpoint override.</param>
    /// <param name="model">Model identifier; defaults to <see cref="DefaultModel"/>.</param>
    /// <param name="httpClient">Optional client for tests.</param>
    public DeepgramBatchProvider(
        string apiKey,
        Uri? endpoint = null,
        string model = DefaultModel,
        HttpClient? httpClient = null)
    {
        _apiKey = apiKey;
        _endpoint = endpoint ?? DefaultEndpoint;
        _model = model;
        _httpClient = httpClient ?? HttpDefaults.Default;
    }

    /// <inheritdoc />
    public async Task<BatchTranscriptResult> TranscribeAsync(
        string audioFilePath,
        TranscriptionConfig config,
        CancellationToken ct = default)
    {
        byte[] audioData;
        try
        {
            audioData = await File.ReadAllBytesAsync(audioFilePath, ct).ConfigureAwait(false);
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            throw new SendFailedException($"Could not read audio file: {ex.Message}", ex);
        }

        var url = BuildUrl(_endpoint, _model, config.Language);
        using var request = new HttpRequestMessage(HttpMethod.Post, url);
        request.Headers.TryAddWithoutValidation("Authorization", $"Token {_apiKey}");

        var content = new ByteArrayContent(audioData);
        content.Headers.ContentType = new MediaTypeHeaderValue("audio/m4a");
        request.Content = content;

        HttpResponseMessage response;
        try
        {
            response = await _httpClient.SendAsync(request, ct).ConfigureAwait(false);
        }
        catch (HttpRequestException ex)
        {
            throw new SendFailedException(ex.Message, ex);
        }

        try
        {
            await HttpStatus.ThrowIfNotSuccessAsync(response, ct).ConfigureAwait(false);
            byte[] body = await response.Content.ReadAsByteArrayAsync(ct).ConfigureAwait(false);
            return Parse(body);
        }
        finally
        {
            response.Dispose();
        }
    }

    /// <summary>
    /// Build the GET URL with query items in fixed insertion order
    /// (<c>model</c>, <c>punctuate</c>, <c>smart_format</c>, optional
    /// <c>language</c>). Skips language when null/empty/<c>"auto"</c>.
    /// </summary>
    internal static Uri BuildUrl(Uri endpoint, string model, string? language)
    {
        var sb = new StringBuilder();
        sb.Append(endpoint.GetLeftPart(UriPartial.Path));
        sb.Append('?');
        sb.Append("model=").Append(Uri.EscapeDataString(model));
        sb.Append('&').Append("punctuate=true");
        sb.Append('&').Append("smart_format=true");
        if (!string.IsNullOrEmpty(language) &&
            !string.Equals(language, "auto", StringComparison.OrdinalIgnoreCase))
        {
            sb.Append('&').Append("language=").Append(Uri.EscapeDataString(language));
        }
        return new Uri(sb.ToString());
    }

    /// <summary>
    /// Parse a Deepgram batch JSON response into a <see cref="BatchTranscriptResult"/>.
    /// Mirrors the Swift cursor-based 5-second word-grouping loop exactly.
    /// </summary>
    internal static BatchTranscriptResult Parse(byte[] data)
    {
        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(data);
        }
        catch (JsonException ex)
        {
            throw new DecodingFailedException(ex.Message, ex);
        }

        using (doc)
        {
            var root = doc.RootElement;

            if (!root.TryGetProperty("results", out var results) ||
                !results.TryGetProperty("channels", out var channels) ||
                channels.ValueKind != JsonValueKind.Array ||
                channels.GetArrayLength() == 0)
            {
                throw new DecodingFailedException("No transcript channels in response.");
            }

            var channel = channels[0];
            if (!channel.TryGetProperty("alternatives", out var alternatives) ||
                alternatives.ValueKind != JsonValueKind.Array ||
                alternatives.GetArrayLength() == 0)
            {
                throw new DecodingFailedException("No alternatives in channel.");
            }

            var alt = alternatives[0];
            if (!alt.TryGetProperty("transcript", out var transcriptEl) ||
                transcriptEl.ValueKind != JsonValueKind.String)
            {
                throw new DecodingFailedException("Missing transcript text.");
            }
            string transcript = transcriptEl.GetString() ?? string.Empty;

            double duration = 0;
            if (root.TryGetProperty("metadata", out var metadata) &&
                metadata.TryGetProperty("duration", out var durEl) &&
                durEl.ValueKind == JsonValueKind.Number)
            {
                duration = durEl.GetDouble();
            }

            string? language = null;
            if (channel.TryGetProperty("detected_language", out var langEl) &&
                langEl.ValueKind == JsonValueKind.String)
            {
                language = langEl.GetString();
            }

            // Materialize words once.
            var words = new List<DgWord>();
            if (alt.TryGetProperty("words", out var wordsEl) &&
                wordsEl.ValueKind == JsonValueKind.Array)
            {
                foreach (var w in wordsEl.EnumerateArray())
                {
                    double start = w.TryGetProperty("start", out var s) ? s.GetDouble() : 0;
                    double end = w.TryGetProperty("end", out var e) ? e.GetDouble() : 0;
                    string? punct = w.TryGetProperty("punctuated_word", out var p) &&
                                    p.ValueKind == JsonValueKind.String
                        ? p.GetString()
                        : null;
                    string word = w.TryGetProperty("word", out var ww) ? ww.GetString() ?? string.Empty : string.Empty;
                    words.Add(new DgWord(start, end, word, punct));
                }
            }

            var segments = new List<TranscriptSegment>();
            if (words.Count == 0)
            {
                segments.Add(new TranscriptSegment(
                    Start: 0,
                    End: duration,
                    Text: transcript,
                    Confidence: 1.0,
                    IsFinal: true,
                    Speaker: null));
            }
            else
            {
                int cursor = 0;
                while (cursor < words.Count)
                {
                    double segmentStart = words[cursor].Start;
                    double cutoff = segmentStart + 5.0;
                    int endIndex = cursor;
                    while (endIndex < words.Count && words[endIndex].End <= cutoff)
                    {
                        endIndex++;
                    }
                    if (endIndex == cursor) endIndex = cursor + 1;

                    var sb = new StringBuilder();
                    for (int i = cursor; i < endIndex; i++)
                    {
                        if (i > cursor) sb.Append(' ');
                        sb.Append(words[i].Punctuated ?? words[i].Word);
                    }
                    double segmentEnd = words[endIndex - 1].End;
                    segments.Add(new TranscriptSegment(
                        Start: segmentStart,
                        End: segmentEnd,
                        Text: sb.ToString(),
                        Confidence: 1.0,
                        IsFinal: true,
                        Speaker: null));
                    cursor = endIndex;
                }
            }

            return new BatchTranscriptResult(language, duration, segments, transcript);
        }
    }

    private readonly record struct DgWord(double Start, double End, string Word, string? Punctuated);
}
